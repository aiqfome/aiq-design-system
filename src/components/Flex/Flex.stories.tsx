import React from 'react'
import { withKnobs, select, text } from '@storybook/addon-knobs'

import { Flex } from './Flex'
import { Text } from '../Text'

export default {
  component: Flex,
  title: 'Flex',
  decorators: [withKnobs]
}

export const basic: React.FC = () => (
  <Flex
    justifyContent={select(
      'justifyContent',
      { FlexStart: 'flex-start', FlexEnd: 'flex-end', Center: 'center' },
      'flex-start'
    )}
    alignItems={select(
      'alignItems',
      { FlexStart: 'flex-start', FlexEnd: 'flex-end', Center: 'center' },
      'flex-start'
    )}
    flexDirection={select(
      'flexDirection',
      { Row: 'row', Column: 'column' },
      'row'
    )}
    backgroundColor={text('backgroundColor', '#fff')}
    variant={select(
      'Variant',
      { Centralized: 'centralized', FullCentralized: 'fullCentralized' },
      'centralized'
    )}
  >
    <Text>Flex!</Text>
  </Flex>
)

export const centralized: React.FC = () => (
  <Flex variant='centralized'>
    <Text>Flex!</Text>
  </Flex>
)

export const fullCentralized: React.FC = () => (
  <Flex variant='fullCentralized'>
    <Text>Flex!</Text>
  </Flex>
)
