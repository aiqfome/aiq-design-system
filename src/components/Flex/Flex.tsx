import React from 'react'
import styled, { css, DefaultTheme } from 'styled-components'
import PropTypes from 'prop-types'
import {
  color,
  ColorProps,
  space,
  SpaceProps,
  layout,
  LayoutProps,
  fontSize,
  FontSizeProps,
  fontWeight,
  FontWeightProps,
  border,
  BorderProps,
  position,
  FlexProps,
  flex,
  PositionProps,
  flexbox,
  FlexboxProps
} from 'styled-system'

export interface Props
  extends SpaceProps,
    DefaultTheme,
    LayoutProps,
    FontSizeProps,
    FontWeightProps,
    PositionProps,
    ColorProps,
    FlexProps,
    FlexboxProps,
    BorderProps {
  variant?: 'auto' | 'centralized' | 'fullCentralized'
  color?: string
  fullHeight?: boolean
  children?: any
  className?: string
  onClick?: (e: any) => void
  onDragOver?: (e: any) => void
  onDragEnter?: (e: any) => void
  onDragLeave?: (e: any) => void
  onDragEnd?: (e: any) => void
  onDrop?: (e: any) => void
  style?: any
}

const flexVariations: { [index: string]: any } = {
  auto: css``,
  centralized: css`
    justify-content: center;
    align-items: center;
  `,
  fullCentralized: css`
    justify-content: center;
    align-items: center;
    height: 100vh;
  `
}

const FlexStyled = styled.div<Props>`
  ${space}
  ${fontSize}
  ${fontWeight}
  ${border}
  ${position}
  ${color}
  ${flexbox}
  ${flex}

  display: flex;
  
  ${({ variant }) => flexVariations[variant || 'auto']}

  ${({ fullHeight }) =>
    !!fullHeight &&
    css`
      height: 100vh;
    `}

  ${layout}
`

export const Flex = React.forwardRef<HTMLInputElement, Props>(
  ({ children, color, ...props }, ref) => {
    return (
      <FlexStyled ref={ref} color={color} {...props}>
        {children}
      </FlexStyled>
    )
  }
)

Flex.displayName = 'Flex'

Flex.propTypes = {
  ref: PropTypes.func,
  color: PropTypes.string,
  onClick: PropTypes.func,
  children: PropTypes.node
}
