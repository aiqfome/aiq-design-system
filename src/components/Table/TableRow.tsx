import React, { ReactNode } from 'react'
import PropTypes from 'prop-types'

import styled, { css } from 'styled-components'

export interface TableRowProps {
  expanded?: boolean
  children: ReactNode
  hoverable?: boolean
  hasAction?: boolean
}

const TableRowStyled = styled.tr<TableRowProps>`
  cursor: default;
  vertical-align: middle;
  border-bottom: 1px solid ${({ theme }) => theme.colors.mediumGrey};

  ${({ hasAction }) =>
    hasAction &&
    css`
      cursor: pointer;
    `}

  ${({ expanded }) =>
    expanded &&
    css`
      cursor: default;
      background-color: ${({ theme }) => theme.colors.lightGrey};
    `}

  ${({ hoverable }) =>
    hoverable &&
    css`
      &:hover {
        background-color: ${({ theme }) => theme.colors.lightGrey};
      }
    `}
`

export const TableRow: React.FC<TableRowProps> = ({
  children,
  hoverable,
  ...props
}) => {
  return (
    <TableRowStyled hoverable={hoverable} {...props}>
      {children}
    </TableRowStyled>
  )
}

TableRow.defaultProps = {
  hoverable: true
}

TableRow.propTypes = {
  expanded: PropTypes.bool,
  children: PropTypes.node,
  hasAction: PropTypes.bool,
  hoverable: PropTypes.bool
}
