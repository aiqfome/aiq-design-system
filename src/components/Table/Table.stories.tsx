import React from 'react'

import { Table } from './Table'
import { Flex } from '../Flex'
import { Badge } from '../Badge'
import { Container } from '../Container'

export default {
  component: Table,
  title: 'Table'
}

const renderBadge = value => {
  return (
    <Badge variant='label' backgroundColor='primary'>
      {value}
    </Badge>
  )
}

const data = [
  {
    firstName: 'Joe',
    lastName: 'Dae',
    age: '20',
    friends: '0',
    place: 'casa da mÃ£e joana',
    job: 'desenvolvedor'
  },
  {
    firstName: 'Tom',
    lastName: 'Tompson',
    age: '40',
    friends: '1000',
    place: 'nÃ£o sabemos com certeza',
    job: 'contador'
  },
  {
    firstName: 'Max',
    lastName: 'Maximun',
    age: '24',
    friends: '200',
    place: 'nÃ£o sei, qualquer lugar',
    job: 'design'
  }
]

const columns = [
  {
    name: 'nome', // text/name of column
    accessor: 'firstName' // used as index in data array
  },
  {
    name: 'sobrenome',
    accessor: 'lastName'
  },
  {
    name: 'idade',
    align: 'center',
    accessor: 'age',
    renderCell: renderBadge // special render to data value (value, rowValues) => newValue
  },
  {
    name: 'amigos',
    align: 'center', // used for text-align ('left', 'center', 'right', 'justify')
    accessor: 'friends'
  },
  {
    name: 'lugar preferido',
    width: '150px', // you can use other size props (maxWidth, minWidth, etc..)
    wrap: true, // used to add (...) and wrap text based on width
    accessor: 'place'
  },
  {
    name: 'cargo',
    align: 'center',
    accessor: 'job'
  }
]

export const Basic: React.FC = () => {
  return (
    <Flex variant='fullCentralized' backgroundColor='#FFF'>
      <Table hoverable={false} data={data} columns={columns} />
    </Flex>
  )
}

export const WithScroll: React.FC = () => {
  return (
    <Flex variant='fullCentralized' backgroundColor='#FFF'>
      <Table scroll='1000px' data={data} columns={columns} />
    </Flex>
  )
}

export const WithExpandedRow: React.FC = () => {
  return (
    <Flex variant='fullCentralized' backgroundColor='#FFF'>
      <Table
        data={data}
        columns={columns}
        expandedRowRender={row => `eu amo ser ${row.job}`}
      />
    </Flex>
  )
}

export const WithClickAction: React.FC = () => {
  return (
    <Flex variant='fullCentralized' backgroundColor='#FFF'>
      <Table
        data={data}
        columns={columns}
        onClickRow={row => console.log('todas as infos da row:', row)}
      />
    </Flex>
  )
}

export const WithContainer: React.FC = () => {
  return (
    <Container title='My Table!'>
      <Flex flex={1} flexDirection='column' m='24px'>
        <Table data={data} columns={columns} />
      </Flex>
    </Container>
  )
}
