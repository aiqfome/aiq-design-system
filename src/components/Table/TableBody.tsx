import React from 'react'
import PropTypes from 'prop-types'

import { Flex } from '../Flex'

export const TableBody: React.FC = ({ children, ...props }) => {
  return (
    <Flex flexDirection='column' {...props}>
      {children}
    </Flex>
  )
}

TableBody.propTypes = {
  children: PropTypes.node
}
