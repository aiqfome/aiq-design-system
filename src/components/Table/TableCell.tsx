import React, { ReactNode } from 'react'
import PropTypes from 'prop-types'

import styled from 'styled-components'

import { Flex } from '../Flex'

export interface TableCellProps {
  flex?: number
  children: ReactNode
}

const TableCellStyled = styled(Flex)`
  padding: 0 10px;
  font-size: medium;
`

export const TableCell: React.FC<TableCellProps> = ({
  children,
  flex,
  ...props
}) => {
  return (
    <TableCellStyled flex={flex} {...props}>
      {children}
    </TableCellStyled>
  )
}

TableCell.propTypes = {
  children: PropTypes.node,
  flex: PropTypes.number
}
