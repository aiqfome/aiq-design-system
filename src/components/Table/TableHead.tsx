import React, { useMemo } from 'react'
import PropTypes from 'prop-types'

import { Flex } from '../Flex'

export const TableHead: React.FC = ({ children, ...props }) => {
  const tableProps = useMemo(() => {
    const prop = {
      borderCollapse: 'collapse',
      borderBottom: '0.5px solid',
      borderColor: 'grey',
      flex: 1,
      justifyContent: 'space-between',
      ...props
    }

    return prop
  }, [props])

  return <Flex {...tableProps}>{children}</Flex>
}

TableHead.propTypes = {
  children: PropTypes.node
}
