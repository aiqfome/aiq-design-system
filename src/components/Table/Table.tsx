import React, { Fragment, useCallback, useEffect, useState } from 'react'
import PropTypes from 'prop-types'
import styled, { css } from 'styled-components'

import { useTable } from 'react-table'
import { space, layout, SpaceProps, LayoutProps } from 'styled-system'

import { TableRow } from './TableRow'
import { TableHead } from './TableHead'
import { TableCell } from './TableCell'

export interface TableProps {
  scroll?: string
  data: Array<any>
  columns: Array<any>
  hoverable?: boolean
  onClickRow?: (record: any) => any
  expandedRowRender?: (record: any) => any
}

export interface FlexProps extends SpaceProps, LayoutProps {
  showScroll: boolean
}

const TableStyled = styled.table<TableProps>`
  border-collapse: collapse;
  width: 100%;
  table-layout: fixed;

  ${({ scroll }) =>
    scroll &&
    css`
      min-width: ${scroll};
    `};
`

const FlexWrapper = styled.div<FlexProps>`
  flex: 1;
  display: block;
  max-width: 100%;
  overflow-y: hidden;
  overflow-x: ${({ showScroll }) => (showScroll ? 'scroll' : 'hidden')};

  &::-webkit-scrollbar {
    height: 5px;
    padding-top: 5px;
  }

  &::-webkit-scrollbar-track {
    background-color: #ebebeb;
    -webkit-border-radius: 10px;
    border-radius: 10px;
  }

  &::-webkit-scrollbar-thumb {
    -webkit-border-radius: 10px;
    border-radius: 10px;
    background: ${({ theme }) => theme.colors.grey};
  }

  &::-webkit-scrollbar-thumb:active,
  &::-webkit-scrollbar-thumb:hover {
    -webkit-border-radius: 10px;
    border-radius: 10px;
    background: ${({ theme }) => theme.colors.darkGrey};
  }

  ${layout}
  ${space}
`

export const Table: React.FC<TableProps> = ({
  scroll,
  hoverable,
  data = [],
  onClickRow,
  columns = [],
  expandedRowRender,
  ...props
}) => {
  const {
    rows,
    prepareRow,
    getTableProps,
    getTableBodyProps,
    columns: tableColumns
  } = useTable({ columns, data })

  const [selectedRows, setSelectedRows] = useState<Array<string>>([])

  const [showScroll, setShowScroll] = useState(false)
  const [ref, setRef] = useState<HTMLDivElement | null>(null)

  useEffect(() => {
    const handleResize = () => {
      if (ref && ref.scrollWidth > ref.clientWidth) {
        setShowScroll(true)
      } else {
        setShowScroll(false)
      }
    }

    if (ref) {
      handleResize()
      window.addEventListener('resize', handleResize)
    }
  }, [ref])

  const getRowAction = useCallback(
    row => {
      if (expandedRowRender) {
        if (selectedRows.includes(row.id)) {
          setSelectedRows(rows => rows.filter(r => r !== row.id))
        } else {
          setSelectedRows(rows => rows.concat([row.id]))
        }
      }

      if (onClickRow) {
        onClickRow(data[row.index])
      }
    },
    [expandedRowRender, data, selectedRows, onClickRow]
  )

  return (
    <FlexWrapper
      showScroll={showScroll}
      ref={el => setRef(el || null)}
      {...props}
    >
      <TableStyled scroll={scroll} {...getTableProps()}>
        <thead>
          <TableRow hoverable={false}>
            {tableColumns.map(column => (
              <TableHead
                {...column}
                key={column.id}
                textAlign={column.align || 'left'}
                {...column.getHeaderProps()}
              >
                {column.name ? column.render('name') : column.render('Header')}
              </TableHead>
            ))}
          </TableRow>
        </thead>

        <tbody {...getTableBodyProps()}>
          {rows.map(row => {
            prepareRow(row)
            return (
              <Fragment key={row.id}>
                <TableRow
                  hoverable={hoverable}
                  onClick={() => getRowAction(row)}
                  hasAction={expandedRowRender || onClickRow}
                  {...row.getRowProps()}
                >
                  {row.cells.map(cell => {
                    return (
                      <TableCell
                        {...cell.column}
                        key={`${row.id}-${cell.value}`}
                        textAlign={cell.column.align || 'left'}
                        {...cell.getCellProps()}
                      >
                        {cell.column.renderCell
                          ? cell.column.renderCell(
                              cell.value,
                              data[cell.row.index]
                            )
                          : cell.render('Cell')}
                      </TableCell>
                    )
                  })}
                </TableRow>

                {expandedRowRender && selectedRows.includes(row.id) && (
                  <TableRow expanded>
                    <TableCell colspan={tableColumns.length}>
                      {expandedRowRender(row.values)}
                    </TableCell>
                  </TableRow>
                )}
              </Fragment>
            )
          })}
        </tbody>
      </TableStyled>
    </FlexWrapper>
  )
}

Table.propTypes = {
  scroll: PropTypes.string,
  hoverable: PropTypes.bool,
  onClickRow: PropTypes.func,
  data: PropTypes.array.isRequired,
  expandedRowRender: PropTypes.func,
  columns: PropTypes.array.isRequired
}
