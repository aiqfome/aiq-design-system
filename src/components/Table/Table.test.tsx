import React from 'react'

import { render } from '../utils/test/render'

import { Flex } from '../Flex'
import { Table } from './Table'

describe('must match with the previous snapshot', () => {
  const data = [
    {
      col1: 'Hello',
      col2: 'World'
    },
    {
      col1: 'react-table',
      col2: 'rocks'
    },
    {
      col1: 'whatever',
      col2: 'you want'
    }
  ]

  const columns = [
    {
      Header: 'Column 1',
      accessor: 'col1'
    },
    {
      Header: 'Column 2',
      accessor: 'col2'
    }
  ]

  test('snapshot renders', () => {
    const component = render(
      <Flex variant='fullCentralized' backgroundColor='#F5F5F5'>
        <Table data={data} columns={columns} />
      </Flex>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
