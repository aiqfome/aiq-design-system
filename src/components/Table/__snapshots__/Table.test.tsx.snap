// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`must match with the previous snapshot snapshot renders 1`] = `
<div
  className="sc-bdnylx izFfAG"
>
  <div
    className="sc-jSFkmK liokeG"
  >
    <table
      className="sc-eCApGN jcZVBO"
      role="table"
    >
      <thead>
        <tr
          className="sc-gtssRu guLcfq"
        >
          <th
            className="sc-dlnjPT gLLEWb"
            colSpan={1}
            id="col1"
            role="columnheader"
            width={150}
          >
            Column 1
          </th>
          <th
            className="sc-dlnjPT gLLEWb"
            colSpan={1}
            id="col2"
            role="columnheader"
            width={150}
          >
            Column 2
          </th>
        </tr>
      </thead>
      <tbody
        role="rowgroup"
      >
        <tr
          className="sc-gtssRu ecqTxl"
          onClick={[Function]}
          role="row"
        >
          <td
            className="sc-hKFyIo cHyMeM"
            id="col1"
            role="cell"
            width={150}
            wrap={0}
          >
            Hello
          </td>
          <td
            className="sc-hKFyIo cHyMeM"
            id="col2"
            role="cell"
            width={150}
            wrap={0}
          >
            World
          </td>
        </tr>
        <tr
          className="sc-gtssRu ecqTxl"
          onClick={[Function]}
          role="row"
        >
          <td
            className="sc-hKFyIo cHyMeM"
            id="col1"
            role="cell"
            width={150}
            wrap={0}
          >
            react-table
          </td>
          <td
            className="sc-hKFyIo cHyMeM"
            id="col2"
            role="cell"
            width={150}
            wrap={0}
          >
            rocks
          </td>
        </tr>
        <tr
          className="sc-gtssRu ecqTxl"
          onClick={[Function]}
          role="row"
        >
          <td
            className="sc-hKFyIo cHyMeM"
            id="col1"
            role="cell"
            width={150}
            wrap={0}
          >
            whatever
          </td>
          <td
            className="sc-hKFyIo cHyMeM"
            id="col2"
            role="cell"
            width={150}
            wrap={0}
          >
            you want
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
